### DISCUSSION TITLE
401 Unauthorized when submitting signed challenge and public certificate

### DISCUSSION THREAD

**Discussion by nicolaslang1** [Digitaler-Impfnachweis/certification-apis#106] [Q&A] [answered]
Hello

we are using KOPS 3.1 to place a SMC-B card and authenticate to the keycloak-cert-auth.dev

On step 4 : "4. Submit Signed challenge and SMCB public certificate" we receive the following error:

Keycloak

We are sorry... 

Certificate & Signature Verification failed

**Comment by norganim** [comment, marked as answer]
Hi,

we have the same problem with KoPS 3.1 via keycloak on Step 4.

- used card: SMC-B (Arzt, Standard)
- challenge is hashed (SHA-256) and signed (RSASSA-PSS)
- SMC-B certificate is C.AUT

**Reply by nicolaslang1** [reply]
Hi,

yes we have the exact same configuration. I also tried other SMCBs. Can your please inform me if you can solve this problem somehow. I will inform you too.

**Reply by norganim** [reply]
Hi,

we solved the problem, open the web interface of KoPS, select QES and here under Aufrufkontext, use the card "SMC-B für Impfnachweis" and it worked for us.

**Reply by nicolaslang1** [reply]
Hi, 

very cool ! :) Thanks for your answer. Could you please send me the details of the "SMC-B für Impfnachweis" Card (i mean the information when you click on the little settings button of the card. Because we don't have it.


![image](https://user-images.githubusercontent.com/64006669/122224278-3e531700-ceb4-11eb-8ba1-31aaec9d12d0.png)

Kind regards 

Nicolas

**Reply by norganim** [reply]
Hi,

stop KoPS server
unpack the zip to user/conf/QES
start KoPS server

Edit by @oliver-steinbrecher : Removed file containing sensitive information.

**Reply by nicolaslang1** [reply]
You are awesome :)! it worked☺️

**Reply by norganim** [reply]
No problem you're welcome

**Reply by nicolaslang1** [reply]
Were you able to create a certificate? We only use KOPS and don't have access to the TI so we can't reach the api https://api.ru.impfnachweis.info/api/certify/v2/issue/hash

**Reply by ManuelB** [reply]
I can confirm both RU and PU are working.

**Reply by nicolaslang1** [reply]
Hello Manuel, but are u able to access the ru api without beeing in the TI ?

**Reply by ManuelB** [reply]
No, I am connected to the RU TI and I can do a Teamviewer session into a doctor's practice.

**Comment by ManuelB** [comment]
Our solution is here:

https://github.com/ere-health/dgc-ps-app

It GPL watching is allowed. Copy'n paste requires a commercial license.

**Comment by ManuelB** [comment]
Ich hab mal unsere Lösung hier detailliert erklärt: https://youtu.be/oLrjF-OISO0 https://github.com/ere-health/dgc-ps-app
